src/meta-srv/src/procedure/region_failover.rs

```diff
@@ -0,0 +1,589 @@
+///                  │         │    │                              │
+///                  └─────────┘    │ Sends "Open Region" request  │
+///                                 │ to the candidate Datanode,   │
+///                                 │ and wait for 30 seconds      │
+///                                 │                              │
+///                                 │ Check Datanode returns       │
+///                                 │                              │
+///                         success ├──────────────────────────────┘
+///                                 │                       failed
+///                       ┌─────────▼──────────┐
+///                       │UpdateRegionMetadata│
+///                       └─────────┬──────────┘
+///                                 │
+///                                 │ Updates the Region
+///                                 │ placement metadata
+///                                 │
+///                        ┌────────▼────────┐
+///                        │RegionFailoverEnd│
+///                        └─────────────────┘
+/// ```
+pub struct RegionFailoverProcedure {
+    node: Node,
+    context: RegionFailoverContext,
+}
+
+impl RegionFailoverProcedure {
+    const TYPE_NAME: &'static str = "metasrv-procedure::RegionFailover";
+
+    pub fn new(failed_region: RegionIdent, context: RegionFailoverContext) -> Self {
+        let state = RegionFailoverStart::new();
+        let node = Node {
+            failed_region,
+            state: Some(Box::new(state)),
+        };
+        Self { node, context }
+    }
+
+    fn from_json(json: &str, context: RegionFailoverContext) -> ProcedureResult<Self> {
+        let node: Node = serde_json::from_str(json).context(FromJsonSnafu)?;
+        Ok(Self { node, context })
+    }
+}
+
+#[async_trait]
+impl Procedure for RegionFailoverProcedure {
+    fn type_name(&self) -> &str {
+        Self::TYPE_NAME
+    }
+
+    async fn execute(&mut self, _ctx: &ProcedureContext) -> ProcedureResult<Status> {
+        if let Some(state) = self.node.state.take() {
+            let next_state = state
+                .next(&self.context, &self.node.failed_region)
+                .await
+                .map_err(|e| {
+                    if matches!(e, Error::RetryLater { .. }) {
+                        ProcedureError::retry_later(e)
+                    } else {
+                        ProcedureError::external(e)
+                    }
+                })?;
+            self.node.state = Some(next_state);
+        }
+        Ok(self
+            .node
+            .state
+            .as_ref()
+            .map(|s| s.status())
+            .unwrap_or(Status::Done))
+    }
+
+    fn dump(&self) -> ProcedureResult<String> {
+        serde_json::to_string(&self.node).context(ToJsonSnafu)
+    }
+
+    fn lock_key(&self) -> LockKey {
+        let region_ident = &self.node.failed_region;
```

Comment:
The lock key may be a vector of `[catalog, schema, table, region]`.
For example, when the table is doing DDL such as adding columns etc. We may want to lock it with `[catalog, schema, table]`, and doesn't want to do failover at this time.

Reply:
`[catalog, schema, table, region]` is too coarse. If you are worrying about conflicting with DDL, I think `["catalog.schema.table"]` is enough. WDYT?

Reply:
`[catalog, schema, table, region]` will be sorted inside the lockkey impl, so it might not lock as expected. For example,  region number "0" will be sorted as the first element in lock keys, effectively making all region 0 in all tables locked.

I think what we really need are globally unified lockkeys formation. I looked over the codes, currently lockkeys are mostly formated with `{catalog}.{schema}.{table}/region-{region_number}`, so I temporarily change my code here.

Besides, in table DDL situation, I think it should acquire all table's regions lockkeys formated as above. In that way, it wouldn't be affected by the sorting mechanism inside the lockkey impl.

Reply:
Maybe we can define some key structs and convert them to the final key string.

Reply:
I have tracked this in the tracking issue #286